web/gui/dashboard/src/domains/chart/components/chart-container.tsx

```diff
@@ -0,0 +1,202 @@
+import { getPortalNodeStyles } from "../utils/get-portal-node-styles"
+
+import { fetchDataAction } from "../actions"
+import { selectChartData, selectChartDetails, selectChartFetchDataParams } from "../selectors"
+
+import { Chart } from "./chart"
+
+const getChartURLOptions = (attributes: Attributes) => {
+  const {
+    appendOptions,
+    overrideOptions,
+  } = attributes
+  let ret = ""
+
+  ret += overrideOptions
+    ? overrideOptions.toString()
+    : chartLibrariesSettings.dygraph.options(attributes)
+
+  if (typeof appendOptions === "string") {
+    ret += `|${encodeURIComponent(appendOptions)}`
+  }
+
+  ret += "|jsonwrap"
+
+  // todo
+  const isForUniqueId = false
+  // always add `nonzero` when it's used to create a chartDataUniqueID
+  // we cannot just remove `nonzero` because of backwards compatibility with old snapshots
+  if (isForUniqueId || window.NETDATA.options.current.eliminate_zero_dimensions) {
+    ret += "|nonzero"
+  }
+
+  return ret
+}
+
+
+export type Props = {
+  attributes: Attributes
+  chartUuid: string
+  portalNode: HTMLElement
+}
+
+export const ChartContainer = ({
+  attributes,
+  chartUuid,
+  portalNode,
+}: Props) => {
+  const chartDetails = useSelector((state: AppStateT) => selectChartDetails(
+    state, { id: chartUuid },
+  ))
+
+  const [shouldFetch, setShouldFetch] = useState<boolean>(true)
+  useInterval(() => {
+    setShouldFetch(true)
+  }, 2000)
+
+
+  // todo local state option
+  const globalPanAndZoom = useSelector(selectGlobalPanAndZoom)
+  const isGlobalPanAndZoomMaster = !!globalPanAndZoom && globalPanAndZoom.masterID === chartUuid
+  const isRemotelyControlled = !!globalPanAndZoom
+    && (!isGlobalPanAndZoomMaster || !!globalPanAndZoom.shouldForceTimeRange)
+  const fetchDataParams = useSelector((state: AppStateT) => selectChartFetchDataParams(
+    state, { id: chartUuid },
+  ))
+
+  // don't send new requests too often (throttle)
+  // corresponds to force_update_at in old dashboard
+  // + 50 is because normal loop only happened there once per 100ms anyway..
+  const globalPanAndZoomThrottled = useThrottle(globalPanAndZoom,
+    window.NETDATA.options.current.pan_and_zoom_delay + 50)
+  useEffect(() => {
+    setShouldFetch(true)
+  }, [globalPanAndZoomThrottled])
+
+  const {
+    after: initialAfter = window.NETDATA.chartDefaults.after,
+    before: initialBefore = window.NETDATA.chartDefaults.before,
+  } = attributes
+
+  const chartSettings = chartLibrariesSettings[attributes.chartLibrary]
+  const { hasLegend } = chartSettings
+
+  const [hasPortalNodeBeenStyled, setHasPortalNodeBeenStyled] = useState<boolean>(false)
+  // todo take width via hook/HOC and put this into useMemo
+  const chartWidth = portalNode.getBoundingClientRect().width
+    - (hasLegend ? 140 : 0)
+
+  /**
+   * fetch data
+   */
+  const chartData = useSelector((state: AppStateT) => selectChartData(state, { id: chartUuid }))
+  const dispatch = useDispatch()
+  useEffect(() => {
```

Comment:
Seems like a pretty big effect to fully comprehend. I'd be afraid to change something here =) Can it be somehow decomposed even more?

Reply:
thanks for input, I'll see what i can do